fvm/evm/precompiles/abi.go

```diff
@@ -0,0 +1,223 @@
+func EncodeUint64(inp uint64, buffer []byte, index int) error {
+	if len(buffer) < index+EncodedUint64Size {
+		return ErrBufferTooSmall
+	}
+	encoded := make([]byte, 8)
+	binary.BigEndian.PutUint64(encoded, inp)
+	copy(buffer[index:index+EncodedUint64Size],
+		gethCommon.LeftPadBytes(encoded, EncodedUint64Size),
+	)
+	return nil
+}
+
+// ReadUint256 reads an address from the buffer at index
+func ReadUint256(buffer []byte, index int) (*big.Int, error) {
+	if len(buffer) < index+EncodedUint256Size {
+		return nil, ErrInputDataTooSmall
+	}
+	// data is expected to be big endian (zero-padded on the left side)
+	return new(big.Int).SetBytes(buffer[index : index+EncodedUint256Size]), nil
+}
+
+// ReadBytes4 reads a 4 byte slice from the buffer at index
+func ReadBytes4(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+Bytes4DataReadSize {
+		return nil, ErrInputDataTooSmall
+	}
+	// fixed-size byte values are zero-padded on the right side.
+	return buffer[index : index+Bytes4DataReadSize], nil
+}
+
+// ReadBytes8 reads a 8 byte slice from the buffer at index
+func ReadBytes8(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+Bytes8DataReadSize {
+		return nil, ErrInputDataTooSmall
+	}
+	// fixed-size byte values are zero-padded on the right side.
+	return buffer[index : index+Bytes8DataReadSize], nil
+}
+
+// ReadBytes32 reads a 32 byte slice from the buffer at index
+func ReadBytes32(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+Bytes32DataReadSize {
+		return nil, ErrInputDataTooSmall
+	}
+	return buffer[index : index+Bytes32DataReadSize], nil
+}
+
+// EncodeBytes32 encodes data into a bytes 32
+func EncodeBytes32(data []byte, buffer []byte, index int) error {
+	if len(data) > EncodedBytes32Size {
+		return ErrDataTooLarge
+	}
+	if len(buffer) < index+EncodedBytes32Size {
+		return ErrBufferTooSmall
+	}
+	copy(buffer[index:index+EncodedBytes32Size],
+		gethCommon.RightPadBytes(data, EncodedBytes32Size),
+	)
+	return nil
+}
+
+// ReadBytes reads a variable length bytes from the buffer
+func ReadBytes(buffer []byte, index int) ([]byte, error) {
+	if len(buffer) < index+EncodedUint64Size {
+		return nil, ErrInputDataTooSmall
+	}
+	// reading offset (we read into uint64) and adjust index
+	offset, err := ReadUint64(buffer, index)
+	if err != nil {
+		return nil, err
+	}
+	index = int(offset)
+	if len(buffer) < index+EncodedUint64Size {
+		return nil, ErrInputDataTooSmall
+	}
+	// reading length of byte slice
+	length, err := ReadUint64(buffer, index)
+	if err != nil {
+		return nil, err
+	}
+	index += EncodedUint64Size
+	if len(buffer) < index+int(length) {
+		return nil, ErrInputDataTooSmall
+	}
+	return buffer[index : index+int(length)], nil
+}
+
+// SizeNeededForBytesEncoding computes the number of bytes needed for bytes encoding
+func SizeNeededForBytesEncoding(data []byte) int {
+	paddedSize := (len(data) / FixedSizeUnitDataReadSize) + FixedSizeUnitDataReadSize
```

Comment:
I'm not sure I correctly understand the intention of this function, but wouldn't padded size be in some case 0 if the read size unit matches the data length, in this case we always add one window

Reply:
Added more docs to describe what we are doing here.